python/pylibcudf/pylibcudf/column.pyx

```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+    #    0b10101101 >> 6 = 0b00000010
+    # 4. Remove all other bits
+    #    0b00000010 & 1 = 0 (=> the 7th element is null)
+    if null_bitmap == NULL:
+        return 1
+    return ((null_bitmap[bit_index >> 3] >> (bit_index & 7)) & 1) != 0
+
+
+cdef inline void _set_item(list out, int64_t i, object obj) except *:
+    # https://docs.python.org/3/c-api/list.html#c.PyList_SetItem
+    Py_INCREF(obj)
+    PyList_SetItem(out, i, obj)
+
+
+cdef list _arrow_to_pylist_impl(type_id dtype, ArrowArray* arr):
+    cdef size_t n = <size_t>arr.length
+    cdef list out = <list>PyList_New(n)
+    if out is None:
+        raise MemoryError("Unable to create new python list")
+
+    if n == 0:
+        return out
+
+    cdef const void** buffers = arr.buffers
+    cdef int64_t offset = arr.offset
+    cdef size_t bit_index
+    cdef int64_t i, N = n
+    cdef const uint8_t* null_bm = NULL
+    cdef const char* numeric_base = NULL
+    cdef size_t item_size = 0
+    cdef const uint8_t* bool_base = NULL
+    cdef const int32_t* str_offsets = NULL
+    cdef const char* str_base = NULL
+    cdef bint is_bool = dtype == type_id.BOOL8
+    cdef bint is_string = dtype == type_id.STRING
+    cdef bint check_valid = arr.null_count != 0
+
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
+
+    if dtype in (
+        type_id.INT8, type_id.INT16, type_id.INT32, type_id.INT64,
+        type_id.UINT8, type_id.UINT16, type_id.UINT32, type_id.UINT64,
+        type_id.FLOAT32, type_id.FLOAT64, type_id.BOOL8, type_id.STRING
+    ):
+        if is_string:
+            if arr.n_buffers < 3 or buffers[1] == NULL or buffers[2] == NULL:
+                for i in range(N):
+                    _set_item(out, i, None)
+                return out
+        else:
+            if arr.n_buffers < 2 or buffers[1] == NULL:
+                for i in range(N):
+                    _set_item(out, i, None)
+                return out
+
+        if is_bool:
+            bool_base = <const uint8_t*>buffers[1]
+        elif is_string:
+            str_base = <const char*>buffers[2]
+            str_offsets = <const int32_t*>buffers[1] + offset
+        else:
+            if dtype == type_id.INT8 or dtype == type_id.UINT8:
+                item_size = sizeof(uint8_t)
+            elif dtype == type_id.INT16 or dtype == type_id.UINT16:
+                item_size = sizeof(uint16_t)
+            elif dtype == type_id.INT32 or dtype == type_id.UINT32:
+                item_size = sizeof(uint32_t)
+            elif dtype == type_id.INT64 or dtype == type_id.UINT64:
+                item_size = sizeof(uint64_t)
+            elif dtype == type_id.FLOAT32:
+                item_size = sizeof(float)
+            else:
+                item_size = sizeof(double)
+            numeric_base = <const char*>buffers[1] + item_size * <size_t>offset
+
+        for i in range(N):
+            bit_index = offset + <int64_t>i
```

Comment:
Minor nit, you don't need the bit index for non-nullable string/integral types. You'd have to duplicate the line twice below but it might be worth saving the computation.

Reply:
Done